### PR TITLE
ddtrace/tracer: sql comment extract implementation

### PR DESCRIPTION
**Pull request by barbinimichael**
### What does this PR do?
Extract trace information from sql queries injected with sql comments containing trace tags. Previously, `SQLCommentCarrier` left `Extract` unimplemented. 

### Motivation
This allows for services such as database proxies to consume trace information propagated from a service that sends SQL queries to it.  

### Describe how to test/QA your changes
The `ddtrace/tracer/sqlcomment_test.go` test suite was updated to include `Extract` test cases. 

### Reviewer's Checklist
- [x] Changed code has unit tests for its functionality.
- [x] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

File: ddtrace/tracer/sqlcomment.go
Comment (alexandre-normand) [inline ddtrace/tracer/sqlcomment.go:232]: Future versions are supposed to be backward compatible so I think it probably makes sense to allow other versions to be parsed without errors.

File: ddtrace/tracer/sqlcomment.go
Comment (alexandre-normand) [inline ddtrace/tracer/sqlcomment.go:230]: Because the trace parent standard has a fixed number of bytes for each component, it might be faster and make more sense to skip the split and just extract the range of chars for each component.

File: ddtrace/tracer/sqlcomment_test.go
Comment (alexandre-normand) [inline ddtrace/tracer/sqlcomment_test.go:68]: Maybe we don't want to support this but it might make sense to have a test case where the query includes a few comments in the `/* comment */` format since there could be more than a tracer-injected comment (if the application is already using comments in their queries).

File: ddtrace/tracer/sqlcomment.go
Comment (barbinimichael) [inline ddtrace/tracer/sqlcomment.go:204]: We also potentially parse the parent service, DB service, parent version, and env. I wasn't sure if / where it belonged in the span context

Comment (alexandre-normand) [reply ddtrace/tracer/sqlcomment.go:204]: I'll let the tracing experts chime in but I think what you're doing is correct by ignoring those. They're applicable to Database monitoring when we need some basic info about the caller and aren't always guaranteed to be able to find it via the trace data.

File: ddtrace/tracer/sqlcomment.go
Comment (alexandre-normand) [inline ddtrace/tracer/sqlcomment.go:197]: The datadog tracer prepends the sqlcomment to the query but the open-telemetry implementation _appends_ it to the end of the query. I think it might make sense to try and find a comment with `traceparent` regardless of its position in the query text.

File: ddtrace/tracer/sqlcomment_test.go
Comment (alexandre-normand) [inline ddtrace/tracer/sqlcomment_test.go:149]: This seems fine as a test case and can't find a fault with it as far as the functionality that it's testing but it might actually be misleading to future readers and make them think it's a typical example of an open-telemetry implementation. Here's an actual example from [the spec](https://google.github.io/sqlcommenter/spec/): 

```
SELECT * FROM FOO
/*action='%2Fparam*d',controller='index,'framework='spring',
traceparent='00-5bd66ef5095369c7b0d1f8f4bd33716a-c532cb4098ac3dd2-01',
tracestate='congo%3Dt61rcWkgMzE%2Crojo%3D00f067aa0ba902b7'*/
```

Comment (barbinimichael) [reply ddtrace/tracer/sqlcomment_test.go]: Thanks for the suggestion. I updated the implementation to better handle these cases and keep the regex relatively simple.

**Comment by barbinimichael** [comment]
I've updated the implementation to make searching for the comment tracing information much more performant. This is where most of the overhead was for this function.

File: ddtrace/tracer/sqlcomment.go
Comment (ajgajg1134) [inline ddtrace/tracer/sqlcomment.go:217]: can we flip this condition to align the happy-path to the left more?

File: ddtrace/tracer/sqlcomment.go
Comment (ajgajg1134) [inline ddtrace/tracer/sqlcomment.go:226]: not important for this PR probably, but is there any desire/plan to support 128-bit traceIDs in these sql comments? CC @katiehockman

File: ddtrace/tracer/sqlcomment.go
Comment (ajgajg1134) [inline ddtrace/tracer/sqlcomment.go:240]: Does there need to be a check to ensure that the `traceparent` is long enough here?

File: ddtrace/tracer/sqlcomment_test.go
Comment (ajgajg1134) [inline ddtrace/tracer/sqlcomment_test.go:131]:
```suggestion
				require.True(t, ok)
```

**Review by ajgajg1134 — CHANGES_REQUESTED** [review]
Overall looks good just a few smaller comments. Only bigger ask is a fuzz test for this extract to ensure we don't miss any possible panics in the complexity

File: ddtrace/tracer/sqlcomment.go
Comment (alexandre-normand) [reply ddtrace/tracer/sqlcomment.go:226]: This might actually be a good idea since since can technically be used to parse otel sql comments that would be 128 bits (as far as I can tell given the [sqlcommenter examples](https://google.github.io/sqlcommenter/spec/#url-encode) from the otel implementations)

File: ddtrace/tracer/sqlcomment.go
Comment (alexandre-normand) [reply ddtrace/tracer/sqlcomment.go:240]: I like that idea too. The current version (that future versions should be backward compatible with) is 55 bytes long (unless I miscounted) so we could check that `len(traceParent) >= 55`.

File: ddtrace/tracer/sqlcomment.go
Comment (barbinimichael) [reply ddtrace/tracer/sqlcomment.go]: I updated the PR to support this as well

**Review by ajgajg1134 — APPROVED** [review]

**Comment by ajgajg1134** [comment]
Do we want to wait to merge until we have an approval from DBM as well?

**Review by alexandre-normand — APPROVED** [review]
Looking good. Nice to see the support for extraction with some use-cases.

**Pull request merged**
